File: apollo_dump/_archived_20260605/trace_practice_smd.js
// trace_practice_smd.js
// 第一步: hook CreateFileA，抓练习场加载原始 SMD 时的调用栈

// --- 先探测可用 API ---
send({t:'probe', msg:'探测 Frida API...'});
send({t:'probe', msg:'typeof Module = ' + typeof Module});
send({t:'probe', msg:'typeof Process = ' + typeof Process});
send({t:'probe', msg:'typeof Interceptor = ' + typeof Interceptor});
send({t:'probe', msg:'typeof Thread = ' + typeof Thread});

// --- 找 CreateFileA 地址 ---
var pCreateFileA = null;

// 方法 1: Process.getModuleByName (你的 apollo_killer_static.js 用过，确认能用)
try {
    var k32 = Process.getModuleByName('kernel32.dll');
    send({t:'probe', msg:'kernel32 base = ' + k32.base});
    // 尝试枚举导出
    if (typeof k32.enumerateExports === 'function') {
        var exps = k32.enumerateExports();
        for (var i = 0; i < exps.length; i++) {
            if (exps[i].name === 'CreateFileA') {
                pCreateFileA = exps[i].address;
                break;
            }
        }
        send({t:'probe', msg:'enumerateExports 找到 CreateFileA = ' + pCreateFileA});
    } else {
        send({t:'probe', msg:'enumerateExports 不可用'});
    }
} catch (e) {
    send({t:'probe', msg:'方法1失败: ' + e});
}

// 方法 2: 尝试 Module.findExportByName
if (!pCreateFileA) {
    try {
        pCreateFileA = Module.findExportByName('kernel32.dll', 'CreateFileA');
        send({t:'probe', msg:'Module.findExportByName = ' + pCreateFileA});
    } catch (e) {
        send({t:'probe', msg:'方法2失败: ' + e});
    }
}

// 方法 3: 尝试 Module.getExportByName
if (!pCreateFileA) {
    try {
        pCreateFileA = Module.getExportByName('kernel32.dll', 'CreateFileA');
        send({t:'probe', msg:'Module.getExportByName = ' + pCreateFileA});
    } catch (e) {
        send({t:'probe', msg:'方法3失败: ' + e});
    }
}

if (!pCreateFileA) {
    send({t:'error', msg:'所有方法都无法找到 CreateFileA，放弃'});
} else {
    send({t:'ready', msg:'CreateFileA @ ' + pCreateFileA + '，开始 hook'});

    var TARGET_FILE = 'i50125461_FN.smd';
    var hitCount = 0;

    Interceptor.attach(pCreateFileA, {
        onEnter: function (args) {
            var path = null;
            try {
                path = args[0].readUtf8String();
            } catch (e) {
                return;
            }

            if (!path || path.indexOf(TARGET_FILE) === -1) return;

            hitCount++;
            send({t:'hit', n:hitCount, file:path, time:new Date().toTimeString()});

            try {
                var bt = Thread.backtrace(this.context, Backtracer.ACCURATE);
                var frames = [];
                for (var i = 0; i < bt.length; i++) {
                    var addr = bt[i];
                    var mod = Process.findModuleByAddress(addr);
                    frames.push({
                        i: i,
                        addr: addr.toString(),
                        mod: mod ? mod.name : '???',
                        offset: mod ? addr.sub(mod.base).toString() : '?'
                    });
                }
                send({t:'backtrace', n:hitCount, frames:frames});
            } catch (e) {
                send({t:'error', msg:'backtrace失败: ' + e});
            }

            if (hitCount >= 3) {
                send({t:'done', msg:'已捕获 3 次，卸载 hook'});
                Interceptor.detachAll();
            }
        }
    });

    send({t:'ready', msg:'就绪，进练习场触发加载...'});
}

File: apollo_dump/_archived_20260605/trace_practice_smd.py
"""
trace_practice_smd.py — 自动附加 + 抓练习场 SMD 调用栈 + 自动保存日志
用法: python trace_practice_smd.py
"""
import sys, os, time, subprocess, json
from datetime import datetime
sys.stdout.reconfigure(encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
LOG_FILE = os.path.join(SCRIPT_DIR, f'trace_smd_{datetime.now().strftime("%Y%m%d_%H%M%S")}.txt')

log_lines = []

def log(msg):
    line = f'[{time.strftime("%H:%M:%S")}] {msg}'
    print(line)
    log_lines.append(line)

def save_log():
    with open(LOG_FILE, 'w', encoding='utf-8') as f:
        f.write('\n'.join(log_lines))
    print(f'\n日志已保存: {LOG_FILE}')

def find_pid():
    try:
        r = subprocess.run(
            ['tasklist', '/fi', 'IMAGENAME eq FreeStyle.exe', '/fo', 'csv', '/nh'],
            capture_output=True, text=True, timeout=10
        )
        for line in r.stdout.strip().split('\n'):
            if 'FreeStyle.exe' in line:
                return int(line.split(',')[1].strip('"'))
    except Exception as e:
        log(f'查找进程失败: {e}')
    return None

def main():
    pid = find_pid()
    if not pid:
        log('FreeStyle.exe 未运行，请先启动游戏并登录')
        return 1
    log(f'PID={pid}')

    import frida
    session = frida.attach(pid)
    log('Frida 已附加')

    js_path = os.path.join(SCRIPT_DIR, 'trace_practice_smd.js')
    with open(js_path, 'r', encoding='utf-8') as f:
        js_code = f.read()

    def on_msg(msg, data):
        if msg['type'] == 'error':
            log(f'[JS错误] {msg.get("description","")} @ line {msg.get("lineNumber","?")}')
            return
        if msg['type'] != 'send':
            return

        p = msg['payload']
        t = p.get('t', '')

        if t == 'probe':
            log(f'  [探测] {p["msg"]}')
        elif t == 'info' or t == 'ready':
            log(f'  {p["msg"]}')
        elif t == 'error':
            log(f'  [错误] {p["msg"]}')
        elif t == 'hit':
            log(f'========== HIT #{p["n"]} ==========')
            log(f'  文件: {p["file"]}')
            log(f'  时间: {p["time"]}')
        elif t == 'backtrace':
            log(f'  调用栈 ({len(p["frames"])} 帧):')
            for f in p['frames']:
                log(f'    [{f["i"]}] {f["addr"]}  {f["mod"]} +{f["offset"]}')
        elif t == 'done':
            log(f'  {p["msg"]}')

    script = session.create_script(js_code)
    script.on('message', on_msg)
    script.load()

    log('进练习场触发加载... (Ctrl+C 退出)\n')

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        pass

    script.unload()
    session.detach()
    log('已断开')
    save_log()
    return 0

if __name__ == '__main__':
    try:
        sys.exit(main())
    except Exception as e:
        log(f'异常: {e}')
        save_log()
